onnxruntime/core/providers/common.h

```diff
@@ -42,4 +43,45 @@ inline float clamp(float v, float lo, float hi) {
   return v;
 }
 
+/**
+Tries to call the given function parallelly, splitted into specified batch(es)
+**/
+template <typename F>
+inline void TryBatchParallelFor(concurrency::ThreadPool* tp, int32_t total, F&& fn, int32_t batch_size = 0) {
+  if (tp != nullptr) {
+    if (batch_size <= 0) {
+      // concurrency::ThreadPool has numThreads worker threads.
+      // the current ParallelFor() implementation will use the calling thread to execute fn(0)
+      // so there are (numThreads + 1) threads to run the tasks.
+      batch_size = tp->NumThreads() + 1;
+    }
+    if (batch_size == total) {
+      tp->ParallelFor(total, fn);
+    } else {
+      tp->ParallelFor(batch_size, [&](int batch_index) {
+        int start = batch_index * total / batch_size;
+        int end = (batch_index + 1) * total / batch_size;
+        for (int i = start; i < end; i++) {
+          fn(i);
+        }
+      });
+    }
+  } else {
+#ifdef USE_OPENMP
+#pragma omp parallel for
+#endif
+    for (int32_t i = 0; i < total; ++i) {
+      fn(i);
+    }
+  }
+}
+
+/**
+Tries to call the given function parallelly
+**/
+template <typename F>
+inline void TryParallelFor(concurrency::ThreadPool* tp, int32_t total, F&& fn) {
```

Comment:
I would've preferred the Try* functions to be part of the ThreadPool class itself as static members since they're related - executing stuff on thread pools. If a non-provider project wants to use this function, she shouldn't need to include provider utils/common which might bring in other provider stuff that I don't need.